dapps/src/handlers/reader.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A chunk-producing io::Read wrapper.
+
+use std::io::{self, Read};
+
+use futures::{self, sink, Sink, Future};
+use futures::sync::mpsc;
+use hyper;
+
+type Sender = mpsc::Sender<Result<hyper::Chunk, hyper::Error>>;
+
+const MAX_CHUNK_SIZE: usize = 32 * 1024;
+
+pub struct Reader<R: io::Read> {
```

Comment:
This structure requires very descriptive documentation. I've spent over 10 minutes looking at it to understand why it's done like that.

And if I understand correctly, `sending` is a sink, which sends data directly to a server client. Is that right?

Reply:
Also, I can see that `hyper::Response` is generic over `Body`, so maybe there is a way to use it with every structure which implements stream of `Chunks`? Or anything that implements `AsRef<u8>`. Anyway, even if this is not possible now, maybe we should just open an issue in hyper repo. Just to avoid unnecessary allocation, when read value is moved to a heap when creating a `Chunk.